### ISSUE TITLE
Meta-issue for std::future functionality and design constraints discussion

### ISSUE THREAD

**Issue by dhollman** [executors/futures#15]
As discussed on the first call, I'm putting together an issue to discuss and collect our thoughts on requirements imposed by and executor-relevant functionality provided by `std::future`.  I'm actually going to break this up into multiple issues, one for each of what I see to be relatively orthogonal discussion points, and then collect references to those issues here.

Partitioning of Issues
======================

One potentially helpful way of partitioning discussion on `std::future` compatibilitiy is as follows:

- Construction, destruction, and assignment: PLACEHOLDER_FOR_ISSUE_LINK
- Blocking interaction with the state (`wait()` and `get()`): PLACEHOLDER_FOR_ISSUE_LINK
- Non-blocking (or at least, not "permanently" blocking) interaction with the state (`valid()`, `wait_for()`, and `wait_until()`): PLACEHOLDER_FOR_ISSUE_LINK
- Sharing (`share()`): PLACEHOLDER_FOR_ISSUE_LINK (also see #11)

(I realize that some aspects of this partitioning reflect the bias from my background. For instance, if error handling is your top priority, `get()` and `wait()` are very different things, and it might not even make sense to talk about them together.  Feel free to discuss.)

On `std::future` compatibility in general
=========================================

There is also the meta-question of how much it makes sense to talk about `std::future` at all, since its definition is so heavily tied to `std::thread` and thus its relevance for generalized executors might be rather weak.  While true, I think there are several constraints imposed by the existing design of `std::future`:

- `std::future` imposes a lower bound for executor-relevant functionality and compatibility with respect to the goal of providing something for executors in IS20.  I know it's a pretty low bar, but it's a useful sanity check in design discussions.
- As a follow-on to the first point, if some aspect of the executor proposal's needs *can* be met by `std::future`, that need should serve as a basis for discussion about how our design *better* meets that need.
- Similarly, I think we need to examine which needs of the executor proposal *cannot* (or cannot adequately) be met by `std::future` and prioritize design avenues that address these needs.
- Any existing aspect of the concrete `std::future` we currently have but don't like or want to change has to be an axis of design rather than a fixed point.  For instance, I don't think we can get away with saying that all types conforming to the `Future` concept have an unconditionally non-blocking destructor; rather, the blocking behavior of the `Future` concept destructor has to be either fixed at the current behavior or an axis across which `Future` concepts vary.  (The required degree of specificity here is up for debate.  Can we simply say nothing about the blocking behavior of the destructor?  If so, what does this mean for the usability of `Future` concept in generic code?)

Thoughts?

**Comment by jaredhoberock** [comment]
> For instance, I don't think we can get away with saying that all types conforming to the Future concept have an unconditionally non-blocking destructor; rather, the blocking behavior of the Future concept destructor has to be either fixed at the current behavior or an axis across which Future concepts vary.

Agreed that we can't require that all `Future`s' destruction behavior be uniform. We already have examples of futures which always block in their destructor (`std::future`s returned by `std::async`) and some which never block in their destructor (`std::future`s returned by `std::promise::get_future`).

I'd like to suggest that we approach these issues similarly to the approach we've taken with executors. Namely, we have designed executors to advertise their guarantees. I don't think we need a full-blown properties approach (I don't expect we will need futures to be as adaptable as we require executors to be), but we do need some mechanism for clients to reason about a future's behavioral guarantees at both compile time and runtime.

**Comment by dhollman** [comment]
> I don't think we need a full-blown properties approach (I don't expect we will need futures to be as adaptable as we require executors to be), but we do need some mechanism for clients to reason about a future's behavioral guarantees at both compile time and runtime.

Can we hijack the properties of executors for some of this? For instance (strawman syntax, only if `f` is not a `SemiFuture`):

```c++
execution::query(f.get_executor(), future_destructor_blocking) == blocking.never
```

**Comment by jaredhoberock** [comment]
> Can we hijack the properties of executors for some of this?

Maybe, but it's really a guarantee the future should provide rather than the executor. As far as I can tell, the only future-related guarantee that an executor ought to provide is to name its associated future type via `executor_future_t`. If we separate concerns this way, then all the other future-related guarantees can be gleaned from the future.

I actually had a more traditional solution in mind. Could we use something like type traits to check for these guarantees?

For example:

    future_result_t<std::future<T>> -> T
    future_destructor_blocking_t<std::future<T>> -> blocking_t::possibly_t
    future_destructor_blocking_t<never_blocking_future<T>> -> blocking_t::never_t

Maybe a path forward would be to attempt to catalog all of these kinds of queries we'd like to perform, and then try to answer the question of what is the best mechanism for implementing them.

**Comment by brycelelbach** [comment]
I have a slight preference for not having meta issues and instead splitting them up into concrete issues; it seems this is already partially done?

**Comment by dhollman** [comment]
It's already done for a couple of them, but I don't care if we want to merge them back here.

**Comment by brycelelbach** [comment]
This is important, but I don't think it's critical for version 1. We shouldn't constrain ourselves to `std::future`s design - which means we should get our design right first, then think about how legacy code (`std::future`) fits in.

**Comment by LeeHowes** [comment]
Note that discussion in #79 has concrete changes we should make here.
